### PR TITLE
Make App Check core availability for all supported platforms

### PR DESCRIPTION
**Pull request by maksymmalyhin**
b/193554541

- App Check SDK availability now matches to Firebase Core availability
- App Attest and DeviceCheck providers availability matches the underlying platform API availability (see [DCDevice](https://developer.apple.com/documentation/devicecheck/dcdevice) and [DCAppAttestService](https://developer.apple.com/documentation/devicecheck/dcappattestservice))

### CODE REVIEW

File: FirebaseAppCheck/Sources/Public/FirebaseAppCheck/FIRAppCheckAvailability.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// Targets where DeviceCheck framework is available to be used in preprocessor conditions.
+#define FIR_DEVICE_CHECK_SUPPORTED_TARGETS TARGET_OS_IOS || TARGET_OS_OSX || TARGET_OS_TV
+
+// `DeviceCheckProvider` availability.
+#define FIR_DEVICE_CHECK_PROVIDER_AVAILABILITY \
+  API_AVAILABLE(ios(11.0), macos(10.15), tvos(11.0)) API_UNAVAILABLE(watchos)
+
+#pragma mark - App Attest
+
+// App Attest availability was extended to macOS and maccatalyst in Xcode 12.5.
```
Comment (ncooke3) [inline FirebaseAppCheck/Sources/Public/FirebaseAppCheck/FIRAppCheckAvailability.h:32]: nit: consider _Catalyst_ or _Mac Catalyst_ over _maccatalyst_

**Review by ncooke3 — APPROVED** [review]
LGTM 👍

File: FirebaseAppCheck/Sources/Public/FirebaseAppCheck/FIRAppCheckAvailability.h (hunk 1 above)
Comment (maksymmalyhin) [reply FirebaseAppCheck/Sources/Public/FirebaseAppCheck/FIRAppCheckAvailability.h]: done

File: FirebaseAppCheck.podspec [hunk 2]
```diff
@@ -36,7 +38,9 @@ Pod::Spec.new do |s|
   ]
   s.public_header_files = base_dir + 'Sources/Public/FirebaseAppCheck/*.h'
 
-  s.framework = 'DeviceCheck'
+  s.ios.framework = 'DeviceCheck'
+  s.osx.framework = 'DeviceCheck'
+  s.tvos.framework = 'DeviceCheck'
```
Comment (ryanwilson) [inline FirebaseAppCheck.podspec:43]: I think we may need to do `weak_frameworks` if we're pushing back support for older versions that this framework isn't available

File: FirebaseAppCheck/Sources/AppAttestProvider/DCAppAttestService+FIRAppAttestService.h [hunk 3]
```diff
@@ -16,21 +16,22 @@
...
 
 #import <DeviceCheck/DeviceCheck.h>
 
 #import "FirebaseAppCheck/Sources/AppAttestProvider/FIRAppAttestService.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
-API_AVAILABLE(ios(14.0))
-API_UNAVAILABLE(macos, tvos, watchos)
+FIR_APP_ATTEST_PROVIDER_AVAILABILITY
```
Comment (ryanwilson) [inline FirebaseAppCheck/Sources/AppAttestProvider/DCAppAttestService+FIRAppAttestService.h:30]: Does this still show appropriately in docs still?

File: FirebaseAppCheck/Tests/Unit/Swift/AppCheckAPITests.swift [hunk 4]
```diff
@@ -160,16 +160,17 @@ final class AppCheckAPITests {
...
 
     // `DeviceCheckProvider` initializer
-    if let app = FirebaseApp.app(), let deviceCheckProvider = DeviceCheckProvider(app: app) {
-      // Get token
-      deviceCheckProvider.getToken { token, error in
-        if let _ /* error */ = error {
-          // ...
-        } else if let _ /* token */ = token {
-          // ...
+    if #available(iOS 11.0, macOS 10.15, macCatalyst 11.0, tvOS 11.0, *) {
```
Comment (ryanwilson) [inline FirebaseAppCheck/Tests/Unit/Swift/AppCheckAPITests.swift:163]: Same re: Catalyst is 13.0+

File: FirebaseAppCheck/Tests/Unit/Core/FIRAppCheckIntegrationTests.m [hunk 5]
```diff
@@ -78,8 +78,11 @@ - (void)setUp {
 
 - (void)tearDown {
   [FIRApp resetApps];
-  // Recover default provider factory.
-  [FIRAppCheck setAppCheckProviderFactory:[[FIRDeviceCheckProviderFactory alloc] init]];
+
+  if (@available(iOS 11.0, macOS 10.15, macCatalyst 11.0, tvOS 11.0, *)) {
```
Comment (ryanwilson) [inline FirebaseAppCheck/Tests/Unit/Core/FIRAppCheckIntegrationTests.m:82]: I think Catalyst is 13.0+ right?

File: FirebaseAppCheck.podspec (hunk 2 above)
Comment (maksymmalyhin) [reply FirebaseAppCheck.podspec:43]: Good point, let me update it.

File: FirebaseAppCheck/Sources/AppAttestProvider/DCAppAttestService+FIRAppAttestService.h (hunk 3 above)
Comment (maksymmalyhin) [reply FirebaseAppCheck/Sources/AppAttestProvider/DCAppAttestService+FIRAppAttestService.h:30]: Hm, it doesn't, but I cannot find availability displayed anywhere in the docs for other APIs... Where should I check?

Comment (maksymmalyhin) [reply FirebaseAppCheck/Sources/AppAttestProvider/DCAppAttestService+FIRAppAttestService.h:30]: The documentation and generated Swift API is not different from ones when `API_AVAILABLE` is added directly.

File: FirebaseAppCheck/Tests/Unit/Swift/AppCheckAPITests.swift (hunk 4 above)
Comment (maksymmalyhin) [reply FirebaseAppCheck/Tests/Unit/Swift/AppCheckAPITests.swift]: updated

File: FirebaseAppCheck.podspec (hunk 2 above)
Comment (maksymmalyhin) [reply FirebaseAppCheck.podspec]: done

**Comment by maksymmalyhin** [comment]
I'm merging it soon, but feel free to left more comments to address in followup PRs.

**Pull request merged**